### DISCUSSION TITLE
FIREBASE PINTEREST: modal form popping up twice

### DISCUSSION THREAD

**Discussion by katherinevfry** [nss-evening-cohort-14/evening-client-side#166] [Help Me!] [answered]
Katy Fry

Have you searched the other help tickets?
I have.

Describe your issue:
When I click "edit pin" button on my pin card, a modal should pop up, but I'm having to click twice to get it to show. However, then two modals show up. I have to close out of the first one to get to the one that slides in behind it. Only the second modal allows me to update the pin. (The data IS being updated in firebase and on the DOM. The issue seems to lie in the modal). It's almost as if when I click the confirm edit button it's firing up another modal, but I've compared the two buttons and they have different IDs. 

List any error messages that appear
N/A

What have you googled?
https://stackoverflow.com/questions/42124403/bootstrap-modal-firing-show-event-twice
https://github.com/cypress-io/cypress/issues/2877
https://www.toptal.com/twitter-bootstrap/the-10-most-common-bootstrap-mistakes

What have you tried? Tell us everything you have tried.
I've checked my modal against the bootstrap docs and it's identical in structure
I compared my modal to my almost amazon modal
I compared my dom event to my almost amazon modal dom event
I double checked to make sure my edit pin button had the correct data target

Links to my code:
Modal: 
https://github.com/katherinevfry/ASSIGNMENT-Firebase-Pinterest/blob/update/src/javascripts/components/formModal.js#L3
Dom Events:
https://github.com/katherinevfry/ASSIGNMENT-Firebase-Pinterest/blob/update/src/javascripts/components/events/domEvents.js#L66

(PS my Javascript data structure is a mess right now. I'm going to restructure it once I get everything working.)

**Comment by drteresavasquez** [comment]
@katherinevfry You are waiting for the promise to finish before the modal loads here: https://github.com/katherinevfry/ASSIGNMENT-Firebase-Pinterest/blob/update/src/javascripts/components/events/domEvents.js#L69 Which is why there is a delay, so if you click the button again, it triggers another modal.

You should instead use the pattern we used in almost amazon to attach the modal to the DOM on click AND THEN fill the insides with the form you want to give it. 

Modal Component: https://github.com/drteresavasquez/IN-CLASS-almost-amazon-template/blob/main/src/javascripts/components/forms/formModal.js

Modal Body: https://github.com/drteresavasquez/IN-CLASS-almost-amazon-template/blob/main/src/javascripts/components/forms/editBookForm.js

Modal initialized and then form loaded on promise response: https://github.com/drteresavasquez/IN-CLASS-almost-amazon-template/blob/main/src/javascripts/events/domEvents.js#L46-L50

Once the user either submits  the modal is toggled to close: https://github.com/drteresavasquez/IN-CLASS-almost-amazon-template/blob/main/src/javascripts/events/domEvents.js#L66

**Comment by katherinevfry** [comment, marked as answer]
SOLUTION: You need to create your modal and form separately. If you combine them into one module, it's going to run the promise each time you click the button to launch the modal, which will launch the modal more than once. You want the modal to launch once and *then* run the promise.

**Comment by drteresavasquez** [comment]
@chiestroud and @hmparsons-sde try doing this line to remove the event listeners on log out. This helps to prevent the multiple modals and extra cards, etc. On logout. https://github.com/drteresavasquez/black-history-trivia/blob/9414a9dd86f2270ef8a4b730b445fb8ceb8eb762/src/javascripts/components/logoutButton.js#L7

**Comment by jrobinson0529** [comment]
So when I fixed this I had to add that line of code ```document.querySelector('body').removeEventListener('click', domClickEvents);```
as well as ```window.location.reload()``` at the end of the ```signMeOut``` function.

**Reply by drteresavasquez** [reply]
You shouldn’t have needed to do that reload if the logout rerenders the app. If you have event listeners anywhere else you would need to remove them as well. The reload would actually make the need to remove the event listeners unnecessary  but we don’t really ever want to app to reload yet in their case, on logout, it could be ok.
